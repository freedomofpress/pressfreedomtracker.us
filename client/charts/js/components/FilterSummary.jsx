import React from 'react'
import classNames from 'classnames'

const symbolMap = {
	"Arrest / Criminal Charge": "arrest",
	"Border Stop": "border_stop",
	"Denial of Access": "denial_of_access",
	"Equipment Search or Seizure": "equipment_search",
	"Assault": "assault",
	"Leak Case": "leak_case",
	"Subpoena / Legal Order": "subpoena",
	"Equipment Damage": "equipment_damage",
	"Prior Restraint": "prior_restraint",
	"Chilling Statement": "chilling_statement",
	"Other Incident": "other_incident",
}

export default function FilterSummary({ serializedFilters }) {
	const categoryFilters = JSON.parse(serializedFilters)

	const idMap = categoryFilters.reduce((acc, val) => ({ ...acc, [val.id]: val.title }), {});
	const allFilters = categoryFilters.flatMap(category => category.filters).reduce(
		(acc, val) => ({...acc, [val.name]: val})
	)

	const displayText = (filterName, value) => {
		const filter = allFilters[filterName]
		if (filter?.type == "bool") {
			return (value === '1') ? filter.present_summary_name : filter.absent_summary_name
		} else {
			return value.replaceAll("_", " ").toLowerCase()
		}
	}

	const searchParams = new URLSearchParams(window.location.search)
	const {
		categories: categoriesStr,
		tags: tagsStr,
		date_lower,
		date_upper,
		...restFilters
	} = Object.fromEntries(searchParams)

	const categories = [...new Set(
		(categoriesStr ? categoriesStr.split(',').map(d => d.trim()) : [])
			.map(category => {
				if (symbolMap[category]) return category;
				if (idMap[category]) return idMap[category];
				return null;
			})
			.filter(d => d)
	)]

	const tags = tagsStr ? tagsStr.split(',').map(d => d.trim()) : []

	const clearFilter = (filterKey, newFilterValue) => {
		const url = new URL(window.location);
		const newSearchParams = new URLSearchParams(window.location.search)

		if (newFilterValue) newSearchParams.set(filterKey, newFilterValue)
		else newSearchParams.delete(filterKey)

		url.search = newSearchParams.toString()
		window.location = url.toString()
	}

	return (
		<ul className="filters-summary">
			{categories.map(category => (
				<li key={category}>
					<button
						onClick={() => clearFilter(
							"categories", categories.filter(d => d !== category).join(',')
						)}
						className="btn btn-tag"
						aria-label={`Removes filter: ${category.toLowerCase()}`}
					>
						{symbolMap[category] && <div className={classNames("category", `category-${symbolMap[category]}`)}></div>}
						<span>{category.toLowerCase()}</span>
						<span className="close-icon" />
					</button>
				</li>
			))}
			{date_lower && (
				<li>
					<button
						onClick={() => clearFilter("date_lower")}
						className="btn btn-tag"
						aria-label={`Removes start date: ${date_lower}`}
					>
						<span>from: {date_lower}</span>
						<span className="close-icon" />
					</button>
				</li>
			)}
			{date_upper && (
				<li>
					<button
						onClick={() => clearFilter("date_upper")}
						className="btn btn-tag"
						aria-label={`Removes end date: ${date_upper}`}
					>
						<span>to: {date_upper}</span>
						<span className="close-icon" />
					</button>
				</li>
			)}
			{tags.map(tag => (
				<li key={tag}>
					<button
						onClick={() => clearFilter(
							"tags", tags.filter(d => d !== tag).join(',')
						)}
						className="btn btn-tag"
						aria-label={`Removes tag: ${tag.toLowerCase()}`}
					>
						<span>{tag.toLowerCase()}</span>
						<span className="close-icon" />
					</button>
				</li>
			))}
			{Object.keys(restFilters).map(filterKey => (
				<li key={restFilters[filterKey]}>
					<button
						onClick={() => clearFilter(filterKey)}
						className="btn btn-tag"
						aria-label={`Removes filter: ${restFilters[filterKey].replaceAll("_", " ").toLowerCase()}`}
					>
						<span>{displayText(filterKey, restFilters[filterKey])}</span>
						<span className="close-icon" />
					</button>
				</li>
			))}
		</ul>
	)
}
