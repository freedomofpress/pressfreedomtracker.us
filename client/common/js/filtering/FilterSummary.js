import React, { PureComponent } from 'react'
import CategoryList from '~/filtering/CategoryList'
import FiltersList from '~/filtering/FiltersList'


class FilterSummary extends PureComponent {
	render() {
		const {
			categories,
			categoriesEnabled,
			changeFiltersMessage,
			filtersExpanded,
			filterValues,
		} = this.props

		const hasFilters = Object.keys(filterValues).length > 0

		if (!hasFilters) {
			return (
				<div className="filters__summary filters__text--dim">
					{changeFiltersMessage || 'No filters applied.'}
				</div>
			)
		}

		if (filtersExpanded) {
			return (
				<div className="filters__summary">
					Filters
				</div>
			)
		}


		return (
			<div className="filters__summary filters__summary--can-compact filters__text--dim">
				Showing <CategoryList categories={categories.filter(({ id }) => categoriesEnabled[id])} />
				{' '}
				<FiltersList
					categories={categories}
					categoriesEnabled={categoriesEnabled}
					filterValues={filterValues}
				/>
			</div>
		)
	}
}


export default FilterSummary
