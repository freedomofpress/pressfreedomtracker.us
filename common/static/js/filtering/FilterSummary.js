import React, { PureComponent } from 'react'
import CategoryList from '~/filtering/CategoryList'
import FiltersList from '~/filtering/FiltersList'


class FilterSummary extends PureComponent {
	render() {
		const {
			filtersExpanded,
			filterValues,
			categoriesEnabled,
			changeFiltersMessage,
		} = this.props

		const hasAnyFilters = (
			Object.keys(filterValues).length > 0 ||
			categoriesEnabled.some(category => {
				// Check if a non-General category has been whitelisted.
				return category.id !== -1 && category.enabled
			})
		)

		if (!hasAnyFilters) {
			return (
				<div className="filters__summary filters__text--dim">
					{changeFiltersMessage || 'No filters applied.'}
				</div>
			)
		}

		if (filtersExpanded) {
			return (
				<div className="filters__summary">
					Filters
				</div>
			)
		}

		const categories = categoriesEnabled.filter(({ enabled }) => enabled)
		const hasFilters = Object.keys(filterValues).length > 0

		return (
			<div className="filters__summary filters__summary--can-compact filters__text--dim">
				Showing <CategoryList categories={categories} />
				{hasFilters && ' '}
				{hasFilters && <FiltersList filterValues={filterValues} />}
			</div>
		)
	}
}


export default FilterSummary
