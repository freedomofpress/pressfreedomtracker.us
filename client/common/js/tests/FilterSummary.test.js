import React from 'react';
import { shallow, mount } from 'enzyme';
import FilterSummary from '~/filtering/FilterSummary';
import {
	CATEGORIES_FILTER,
	CATEGORIES_ENABLED,
    FILTER_VALUES,
	EMPTY_CATEGORIES_ENABLED,
    EMPTY_FILTER_VALUES,
    FILTERS_EXPANDED
} from '~/tests/factories/CategoriesFilterFactory';

describe('FilterSummary', () => {
	let categories;
	let categoriesEnabled;
    let filterValues;
    let filtersExpanded;

    beforeEach(() => {
        categories = CATEGORIES_FILTER;
        categoriesEnabled = CATEGORIES_ENABLED;
        filterValues = FILTER_VALUES;
        filtersExpanded = FILTERS_EXPANDED;
    });

    afterEach(() => {
        categories = null;
        categoriesEnabled = null;
        filterValues = null;
        filtersExpanded = null;
    });

    it('should render without error in "debug" mode', () => {
        const component = shallow(
            <FilterSummary
                categories={categories}
                categoriesEnabled={categoriesEnabled}
                filterValues={filterValues}
            />
        );

        expect(component.debug()).toMatchSnapshot();
	});

	it('should render FiltersSummary properly with no filters', () => {
		const component = mount(
            <FilterSummary
                categories={categories}
                categoriesEnabled={EMPTY_CATEGORIES_ENABLED}
                filterValues={EMPTY_FILTER_VALUES}
            />
        );
        const filterSummaryText = component.find('.filters__summary.filters__text--dim');
        expect(filterSummaryText.text()).toEqual('No filters applied.');
    });
    
    it('should render FiltersSummary properly with filtersExpanded', () => {
		const component = mount(
            <FilterSummary
                categories={categories}
                categoriesEnabled={categoriesEnabled}
                filtersExpanded={filtersExpanded}
                filterValues={EMPTY_FILTER_VALUES}
            />
        );
        const filterSummaryText = component.find('.filters__summary');
        expect(filterSummaryText.text()).toEqual('Filters');
	});

	it('should show FilterSummary properly according to filters', () => {
		const component = mount(
            <FilterSummary
                categories={categories}
                categoriesEnabled={categoriesEnabled}
                filterValues={filterValues}
            />
		);
		const filterSummary = component.find('div.filters__summary');
		// 1 Category List, 1 Filter List = 2 List
		expect(filterSummary.find('ul.filters__summary-list').length).toEqual(2);
		expect(filterSummary.find('div.filters__text--dim').text()).toEqual(
			'Showing Leak Case Charged under espionage act?: Yes'
		);
	});
});