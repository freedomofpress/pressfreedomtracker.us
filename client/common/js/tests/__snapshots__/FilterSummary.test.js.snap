// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterSummary should render without error in "debug" mode 1`] = `
"<div className=\\"filters__summary filters__summary--can-compact filters__text--dim\\">
  Showing 
  <CategoryList categories={{...}} />
   
  <FiltersList categories={{...}} filterValues={{...}} />
</div>"
`;
